Refactor VideoMediaSampleRenderer to support enqueuing decoded samples

jernoble opened the pull request:
#### b2e420a36ecb45db729eddb8dd38ec1d26b1e439
<pre>
Refactor VideoMediaSampleRenderer to support enqueuing decoded samples
<a href="https://bugs.webkit.org/show_bug.cgi?id=281497">https://bugs.webkit.org/show_bug.cgi?id=281497</a>
<a href="https://rdar.apple.com/137964106">rdar://137964106</a>

Reviewed by Jean-Yves Avenard.

Currently, WebCoreDecompressionSession maintains a queue of decoded images. However,
when WebCoreDecompressionSession is used in a mode where those decoded images are
delivered to its clients, the queue is bypassed. VideoMediaSampleRenderer is a more
natural place for such a cue to belong, and eventually, all queueing behavior should
be migrated from WebCoreDecompressionSession to VideoMediaSampleRenderer. As a first
step, expose the promise-based WebCoreDecompressionSession::enqueueSample() function,
and expose a static method to create a CoreMedia buffer queue. Use these primitives
inside VideoMediaSampleRenderer to enqueue decoded images for retrieval by clients
later.

In this mode of operation, a VideoMediaSampleRenderer will feed decoded images to
its associated AVSampleBufferDisplayLayer or AVSampleBufferVideoRenderer, and store
those images in a queue for retrieval via copyDisplayedPixelBuffer(). This requires
a CMTimebaseRef, and to ensure the main thread doesn&apos;t excessively block these
operations, move most operations to a WorkQueue.

In the future, VideoMediaSampleRenderer can be extended to support having multiple
AVSBDLs or AVSBVRs simultaneously, allowing reconfigurations of layers or renderers
without having to tear down and re-build decoder pipelines.

* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::enqueueSampleBuffer):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::enqueueSample):
(WebCore::MediaPlayerPrivateWebM::setVideoRenderer):
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
(WebCore::VideoMediaSampleRenderer::prefersDecompressionSession):
(WebCore::VideoMediaSampleRenderer::timebase const):
(WebCore::VideoMediaSampleRenderer::setResourceOwner): Deleted.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::VideoMediaSampleRenderer):
(WebCore::VideoMediaSampleRenderer::~VideoMediaSampleRenderer):
(WebCore::VideoMediaSampleRenderer::isReadyForMoreMediaData const):
(WebCore::VideoMediaSampleRenderer::stopRequestingMediaData):
(WebCore::VideoMediaSampleRenderer::setPrefersDecompressionSession):
(WebCore::VideoMediaSampleRenderer::setPrefersDecompressionSessionInternal):
(WebCore::VideoMediaSampleRenderer::setTimebase):
(WebCore::VideoMediaSampleRenderer::setTimebaseInternal):
(WebCore::VideoMediaSampleRenderer::enqueueSample):
(WebCore::VideoMediaSampleRenderer::initializeDecompressionSession):
(WebCore::VideoMediaSampleRenderer::decodedFrameAvailable):
(WebCore::VideoMediaSampleRenderer::purgeDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::ensureDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::flush):
(WebCore::VideoMediaSampleRenderer::resetReadyForMoreSample):
(WebCore::VideoMediaSampleRenderer::expectMinimumUpcomingSampleBufferPresentationTime):
(WebCore::VideoMediaSampleRenderer::copyDisplayedPixelBuffer):
(WebCore::VideoMediaSampleRenderer::totalVideoFrames const):
(WebCore::VideoMediaSampleRenderer::droppedVideoFrames const):
(WebCore::VideoMediaSampleRenderer::corruptedVideoFrames const):
(WebCore::VideoMediaSampleRenderer::totalFrameDelay const):
(WebCore::VideoMediaSampleRenderer::setResourceOwner):
(WebCore::VideoMediaSampleRenderer::copyDisplayedPixelBuffer const): Deleted.
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.h:
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::createBufferQueue):
(WebCore::WebCoreDecompressionSession::enqueueSample):

Canonical link: <a href="https://commits.webkit.org/285965@main">https://commits.webkit.org/285965@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/38dbc078aef11de7f6aed2c000201ab9a65bf635

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74261 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53690 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27072 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78638 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25499 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76378 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62823 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1475 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58375 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16707 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77328 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48536 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63881 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38784 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45478 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21376 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23832 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66926 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21723 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80156 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1578 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/897 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66669 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1723 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63899 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65945 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16383 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9886 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8041 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1542 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4330 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1571 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1559 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1578 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -47,97 +48,226 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
+        setTimebase(nullptr);
+
+        [m_renderer flush];
+        [m_renderer stopRequestingMediaData];
```
renderer are main thread only, should probably do that in the main thread

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -47,97 +48,226 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
-    [m_renderer stopRequestingMediaData];
-    if (m_decompressionSession)
-        m_decompressionSession->invalidate();
+    m_workQueue->dispatchSync([&] {
```
rather than a dispatchSync we could do `m_workQueue->dispatch([decompressionSession = std::exchange(m_decompressionSession, { })] { decompressionSession->invalidata(); });`

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -47,97 +48,226 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
 {
-    return (!m_decompressionSession || m_decompressionSession->isReadyForMoreMediaData()) && [m_renderer isReadyForMoreMediaData];
+    bool ready = false;
+    m_workQueue->dispatchSync([&] {
```
we should avoid dispatchSync if possible. decompressionSession was made to be thread-safe earlier, that should continue.

jernoble replied:
isReadyForMoreMediaData() is definitely not thread-safe.

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -47,97 +48,226 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
-    [m_renderer stopRequestingMediaData];
+    m_workQueue->dispatch([weakThis = ThreadSafeWeakPtr { *this }] {
+        if (RefPtr protectedThis = weakThis.get())
+            [protectedThis->m_renderer stopRequestingMediaData];
```
we are moving from a AVSBDL being main thread only to being accessed concurrently.

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -47,97 +48,226 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
-    m_decompressionSession->decodedFrameWhenAvailable([weakThis = ThreadSafeWeakPtr { *this }](RetainPtr<CMSampleBufferRef>&& sample) {
-        if (RefPtr protectedThis = weakThis.get())
-            [protectedThis->m_renderer enqueueSampleBuffer:sample.get()];
+    m_workQueue->dispatchSync([&] {
```
I think it should be up to the decompressionSession to handle its threading operation, and continue having the VideoMediaSampleRenderer being a main thread object for now.

WebCoreDecompressionSession is already designed to use its own WorkQueue and all operations in parallel. There's no advantage in moving the WebCoreDecompressionSession in yet another WorkQueue.

jernoble replied:
The advantage is not having frames hit the main thread after decoding.

jyavenard replied:
An alternative would be to keep the m_decompressionSession as-is and not worry on which thread it leaves on (it's thread-safe) and only have the m_renderer on its own thread instead . So that any `m_decompressionSession->methodWithPromise()->whenSettled(m_workQueue, [] { do_work_with_avsbdl }`

This removes the need for the sync dispatch

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -46,150 +47,454 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
+
+    if (m_decompressionSession) {
         m_decompressionSession->invalidate();
+        m_decompressionSession = nullptr;
```
not really needed in the destructor

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -46,150 +47,454 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
+
+void VideoMediaSampleRenderer::maybeBecomeReadyForMoreMediaData()
+{
+    if (![renderer() isReadyForMoreMediaData])
```
can we add a isMainThread() assertion here?

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -46,150 +47,460 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
+    CMTime currentTime = PAL::CMTimebaseGetTime(m_timebase.get());
+    CMTime nextPurgeTime = PAL::kCMTimeInvalid;
+
+    while (auto nextSample = (CMSampleBufferRef)const_cast<void*>(PAL::CMBufferQueueGetHead(m_decodedSampleQueue.get()))) {
```
RetainPtr

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -46,150 +47,460 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
+        }
+
+        RetainPtr sampleToBePurged = adoptCF(PAL::CMBufferQueueDequeueAndRetain(m_decodedSampleQueue.get()));
+        sampleToBePurged = nil;
```
unneeded, sampleToBePurged will go out of scope

jernoble replied:
Doing so will result in a "unused variable" warning.

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -46,150 +47,460 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
+    assertIsCurrent(m_workQueue);
+    if (!m_compressedSampleQueue)
+        m_compressedSampleQueue = WebCoreDecompressionSession::createBufferQueue();
+    return m_compressedSampleQueue.get();
```
should be returning RetainPtr<CMBufferQueueRef> to align with new coding safety style

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -46,150 +47,460 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
+    assertIsCurrent(m_workQueue);
+    if (!m_decodedSampleQueue)
+        m_decodedSampleQueue = WebCoreDecompressionSession::createBufferQueue();
+    return m_decodedSampleQueue.get();
```
same comment as above

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -46,150 +47,460 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
     if (m_decompressionSession)
         m_decompressionSession->flush();
+
+    m_workQueue->dispatch([weakThis = ThreadSafeWeakPtr { *this }] {
```
this needs to be made `mutable` for the move below to work

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -46,150 +47,460 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
+        protectedThis->flushCompressedSampleQueue();
+        protectedThis->flushDecodedSampleQueue();
+
+        callOnMainThread([weakThis = WTFMove(weakThis)] {
```
that will no doubt give a compilation error, as `weakThis` is const here

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -46,150 +47,460 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
+{
+    if (m_decompressionSession)
+        return m_decompressionSession->totalFrameDelay();
+    return MediaTime::createWithDouble([renderer() videoPerformanceMetrics].totalFrameDelay);
```
should we add the two values if we have a decompression ses

jernoble replied:
You can't delay a decompressed video frame either.

jyavenard commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -46,150 +47,460 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
+{
+    if (m_decompressionSession)
+        return m_decompressionSession->droppedVideoFrames();
+    return [renderer() videoPerformanceMetrics].numberOfDroppedVideoFrames;
```
I think we should be adding the two

jernoble replied:
You can't drop a uncompressed video frame; it must be displayed.

jyavenard reviewed (approved):
LGTM with nits addressed

jernoble committed:
Refactor VideoMediaSampleRenderer to support enqueuing decoded samples
https://bugs.webkit.org/show_bug.cgi?id=281497
rdar://137964106

Reviewed by Jean-Yves Avenard.

Currently, WebCoreDecompressionSession maintains a queue of decoded images. However,
when WebCoreDecompressionSession is used in a mode where those decoded images are
delivered to its clients, the queue is bypassed. VideoMediaSampleRenderer is a more
natural place for such a cue to belong, and eventually, all queueing behavior should
be migrated from WebCoreDecompressionSession to VideoMediaSampleRenderer. As a first
step, expose the promise-based WebCoreDecompressionSession::enqueueSample() function,
and expose a static method to create a CoreMedia buffer queue. Use these primitives
inside VideoMediaSampleRenderer to enqueue decoded images for retrieval by clients
later.

In this mode of operation, a VideoMediaSampleRenderer will feed decoded images to
its associated AVSampleBufferDisplayLayer or AVSampleBufferVideoRenderer, and store
those images in a queue for retrieval via copyDisplayedPixelBuffer(). This requires
a CMTimebaseRef, and to ensure the main thread doesn't excessively block these
operations, move most operations to a WorkQueue.

In the future, VideoMediaSampleRenderer can be extended to support having multiple
AVSBDLs or AVSBVRs simultaneously, allowing reconfigurations of layers or renderers
without having to tear down and re-build decoder pipelines.

* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::enqueueSampleBuffer):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::enqueueSample):
(WebCore::MediaPlayerPrivateWebM::setVideoRenderer):
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
(WebCore::VideoMediaSampleRenderer::prefersDecompressionSession):
(WebCore::VideoMediaSampleRenderer::timebase const):
(WebCore::VideoMediaSampleRenderer::setResourceOwner): Deleted.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::VideoMediaSampleRenderer):
(WebCore::VideoMediaSampleRenderer::~VideoMediaSampleRenderer):
(WebCore::VideoMediaSampleRenderer::isReadyForMoreMediaData const):
(WebCore::VideoMediaSampleRenderer::stopRequestingMediaData):
(WebCore::VideoMediaSampleRenderer::setPrefersDecompressionSession):
(WebCore::VideoMediaSampleRenderer::setPrefersDecompressionSessionInternal):
(WebCore::VideoMediaSampleRenderer::setTimebase):
(WebCore::VideoMediaSampleRenderer::setTimebaseInternal):
(WebCore::VideoMediaSampleRenderer::enqueueSample):
(WebCore::VideoMediaSampleRenderer::initializeDecompressionSession):
(WebCore::VideoMediaSampleRenderer::decodedFrameAvailable):
(WebCore::VideoMediaSampleRenderer::purgeDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::ensureDecodedSampleQueue):
(WebCore::VideoMediaSampleRenderer::flush):
(WebCore::VideoMediaSampleRenderer::resetReadyForMoreSample):
(WebCore::VideoMediaSampleRenderer::expectMinimumUpcomingSampleBufferPresentationTime):
(WebCore::VideoMediaSampleRenderer::copyDisplayedPixelBuffer):
(WebCore::VideoMediaSampleRenderer::totalVideoFrames const):
(WebCore::VideoMediaSampleRenderer::droppedVideoFrames const):
(WebCore::VideoMediaSampleRenderer::corruptedVideoFrames const):
(WebCore::VideoMediaSampleRenderer::totalFrameDelay const):
(WebCore::VideoMediaSampleRenderer::setResourceOwner):
(WebCore::VideoMediaSampleRenderer::copyDisplayedPixelBuffer const): Deleted.
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.h:
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::createBufferQueue):
(WebCore::WebCoreDecompressionSession::enqueueSample):

Canonical link: https://commits.webkit.org/285965@main
